### DISCUSSION TITLE
read meter data

### DISCUSSION THREAD

**Discussion by pfolcato** [addiejanssen/domoticz-solaredge-modbustcp-plugin#13] [Q&A] [answered]
hi, how can I import the data of Meter in Domoticz?
When I execute the Solaredge_Modbus example.py code I get this result.... but I don't know python.
The meter is connected in Export+Import mode so when I'm able to read "power" I know everything (home consumption... export... import)  
in this example I'm importing 1511 Watt 

"meters": {
        "Meter1": {
            "c_manufacturer": "SolarEdge",
            "c_model": "SE-MTR-3Y-400V-A",
            "c_option": "Export+Import",
            "c_version": "74",
            "c_serialnumber": "xxxxxxxx",
            "c_deviceaddress": 2,
            "c_sunspec_did": 203,
            "current": 68,
            "p1_current": 68,
            "p2_current": 0,
            "p3_current": 0,
            "current_scale": -1,
            "voltage_ln": 24297,
            "p1n_voltage": 24297,
            "p2n_voltage": 41,
            "p3n_voltage": 0,
            "voltage_ll": 0,
            "p12_voltage": 0,
            "p23_voltage": 0,
            "p31_voltage": 0,
            "voltage_scale": -2,
            "frequency": 5000,
            "frequency_scale": -2,
            "power": -1511,
            "p1_power": -1511,
            "p2_power": 0,
            "p3_power": 0,
            "power_scale": 0,
            "power_apparent": 1540,
            "p1_power_apparent": 1539,
            "p2_power_apparent": 0,
            "p3_power_apparent": 0,
            "power_apparent_scale": 0,
            "power_reactive": -295,
            "p1_power_reactive": -295,
            "p2_power_reactive": 0,
            "p3_power_reactive": 0,
            "power_reactive_scale": 0,
            "power_factor": 9736,
            "p1_power_factor": 9210,
            "p2_power_factor": 10000,
            "p3_power_factor": 10000,
            "power_factor_scale": -2,
            "export_energy_active": 21,
            "p1_export_energy_active": 21,
            "p2_export_energy_active": 0,
            "p3_export_energy_active": 0,
            "import_energy_active": 12237,
            "p1_import_energy_active": 12237,
            "p2_import_energy_active": 0,
            "p3_import_energy_active": 0,
            "energy_active_scale": 0,
            "export_energy_apparent": 0,
            "p1_export_energy_apparent": 0,
            "p2_export_energy_apparent": 0,
            "p3_export_energy_apparent": 0,
            "import_energy_apparent": 0,
            "p1_import_energy_apparent": 0,
            "p2_import_energy_apparent": 0,
            "p3_import_energy_apparent": 0,
            "energy_apparent_scale": -32768,
            "import_energy_reactive_q1": 0,
            "p1_import_energy_reactive_q1": 0,
            "p2_import_energy_reactive_q1": 0,
            "p3_import_energy_reactive_q1": 0,
            "import_energy_reactive_q2": 0,
            "p1_import_energy_reactive_q2": 0,
            "p2_import_energy_reactive_q2": 0,
            "p3_import_energy_reactive_q2": 0,
            "export_energy_reactive_q3": 0,
            "p1_export_energy_reactive_q3": 0,
            "p2_export_energy_reactive_q3": 0,
            "p3_export_energy_reactive_q3": 0,
            "export_energy_reactive_q4": 0,
            "p1_export_energy_reactive_q4": 0,
            "p2_export_energy_reactive_q4": 0,
            "p3_export_energy_reactive_q4": 0,
            "energy_reactive_scale": -32768
        }
    },
    "batteries": {}
}

**Comment by addiejanssen** [comment]
Hi pfolcato,

Can you provide a bit more background?
Did you install the plugin already?
Was the plugin able to communicate with your inverter (it will report that in the log)?
Did the plugin install the devices in Domoticz?
When the plugin is failing, can you enable debugging and share the Domoticz log entries?

**Comment by pfolcato** [comment]
hi, the plugin work very well! everything is working, but I need to read another value from inverter.
I have smartmeter connected to Inverter by rs485 and I want to read the "power" table reported in previus post as "power": -1511,
Your plugin read only Inverter data but no smartmeter or battery.
For read smarmeter value I need to add at your plugin this example from sunspec_monitor

>>> inverter.meters()
    {
        'Meter1': Meter1(10.0.0.123:1502, connectionType.TCP: timeout=1, retries=3, unit=0x1)
    }

    >>> meter1 = inverter.meters()["Meter1"]
    >>> meter1
    Meter1(10.0.0.123:1502, connectionType.TCP: timeout=1, retries=3, unit=0x1)

    >>> meter1.read_all()
    {
        'c_manufacturer': 'SolarEdge',
        'c_model': 'PRO380-Mod',
        'c_option': 'Export+Import',
        'c_version': '2.19',
        'c_serialnumber': '12312332',
        'c_deviceaddress': 1,
        'c_sunspec_did': 203,
        'current': -13,
        ...
    }
and use the value "power" as another device. 
But I'm not a python programmer.... :(

**Comment by addiejanssen** [comment]
OK, now I get your point.

The solaredge_modbus library that is used by the plugin has been updated to support inverters, meters and batteries some time ago. However, the plugin is using an approach before those devices were supported.
To do what you are looking for, would mean a pretty big change to the plugin.
Since I do not have access to the hardware myself, testing it would become a challenge.

I will put this on the "requested features" list, but I can not guarantee that this functionality will be added...

**Comment by pfolcato** [comment]
I have to thank you Addiejanssen!
I'm not a programmer but thanks to your plugin i added some code to read the solaredge meter, change the values ​​that are called with the same keys in the inverter and then merge my dictionary variable with your dictionary variable that your plugin processes to assign the read values to the idx of domoticz.
It's not very elegant but it works ;)
If anyone is interested in this gorgeous plugin that allows to create devices in domoticz automatically, read both the inverter and the SolarEdge meter via ModbusTCP, I am attaching the modified plugin.
Thank you again Addiejanssen!
[plugin.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/5611246/plugin.zip)

**Comment by gropi75** [comment]
Great work t both of You.
The plugin works well, I copied the modified part to the 3phase section and it is running well.
![grafik](https://user-images.githubusercontent.com/62841990/101954909-33fe4a80-3bfd-11eb-9511-6ecbcd326ae1.png)

**Comment by pfolcato** [comment]
hello gropi75 I'm glad you used my small part too! the greatest work was done by the owner of this wonderful script!

**Comment by gropi75** [comment]
One update: in case of a 3phase inverter the power will be retured from the meter for each phase and in total. So there is 1 change in the section 3ph needed: use for power just `power` instead of `p1_power`


[plugin.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/5682666/plugin.zip)

**Comment by pfolcato** [comment, marked as answer]
Same plugin for 2 phase inverter, I have added the battery status
[plugin.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/7807773/plugin.zip)

**Comment by vinchz31** [comment]
Hello ropi75, hello pfolcato,
I am looking for a solution to retrieve via Modbus the metrics I find on the SolarEdege site and that I manage to retrieve via the JSON API. In particular, the information provided by the meter: total consumption of the house and imported or exported energy. 
I posted a question in the forum about this ( https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/discussions/53 ) but I think your development answers exactly what I am looking for.
So I tried to replace the initially installed plugin.py file (which works fine but does not return the consumption values) by the different versions you propose but none of them works. Either I don't have any devices being added or the devices are being added fine but all the values remain at 0 with the message you will find below.
Would you be able to help me to get the plugin working?
For information, my inverter is a SE6000H single phase.
Thank you very much for your help!

**Comment by pfolcato** [comment]
I make this modified version of plugin that calculate inside the values that you need (self consumption, purchased energy, energy sold and the power too... I create separate device for better work with Dashticz) but if you don't have the battery you must remove all the parts of "battery" inside my modified script.
Or you can try to use my modified version of plugin only with "meter section added" attached to my post on Nov 28, 2020 
Let me know if it works
[plugin_ok.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/10307422/plugin_ok.zip)

**Reply by vinchz31** [reply]
Thanks a lot for your quick answer! :-)
I just tried the attached file but I don't have any devices that appear in domoticz and I get the following errors in the domoticz log.
I just have :
- removed the previous SolarEdge hardware in domoticz ;
- stopped domoticz
- replaced the plugin.py file with the one you gave me
- restarted domoticz
- added a SolarEdge Modbus TCP type hardware again
Did I do something wrong?
Thanks a lot for your help!


`

**Comment by pfolcato** [comment]
hi, 
Be sure that solaredge_modbus library are correctly installed! (write this to install "pip3 install solaredge_modbus")
Follow the installation instruction, after that, if everything works ok,  replace my plugin.py in the same plugin folder. (If you don't have the battery my plugin doesn't work, you need to remove all parts in plugin.py related to battery)
restart domoticz (sudo service domoticz restart)

Be sure of the connecting parameters (port and ip of inverter) port 11502 seems strange... may be 502 or 1502 but you know your inverter...

**Comment by vinchz31** [comment]
Thanks !
I had initially set the port to 11502 because it was what was indicated on the domoticz wiki (https://www.domoticz.com/wiki/SolarEdge) and I configured the inverter on this port. 

So I think that everything is correctly installed because the module works perfectly well with the initial installation (I get all the information from the inverter, except the ones about the consumption ;-) ). The command "pip3 install solaredge_modbus" tells me that everything is ok.

It's only when I replace the plugin.py by the new one (present in the plugin_Ok.zip file) that it doesn't work anymore (no more device returned) and that errors are returned. If I put back the old plugin, it works again...

I tried to remove all the battery related parts but without success, I get the same result unfortunately...

**Comment by pfolcato** [comment]
ok try with this
[plugin.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/10308416/plugin.zip)

**Comment by pfolcato** [comment]
if it's ok try with this
[plugin_consumption_no_battery.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/10308434/plugin_consumption_no_battery.zip)

**Comment by vinchz31** [comment]
Thanks pfolcato to take time for me !

Both the two version add devices.
The first one add some devices (but not all) but no value is returned by the inverter except for the "current" value
![Inverter1](https://user-images.githubusercontent.com/46422504/209675625-a02d35fc-52df-4245-85c5-f70e256078cc.jpg)

With the second version, I have many more metrics that appear but the values also remain at 0 except the current
![Inverter2](https://user-images.githubusercontent.com/46422504/209678449-9b784507-3c2c-42f7-ada1-46bd8375e71d.jpg)

**Comment by pfolcato** [comment]
try to wait 5 minuts and if nothing work again, put the loglevel to extra and debug true... end paste the result here...

**Comment by vinchz31** [comment]
Nothing append after 5mn.

**Here are the logs in standard mode :**


**You'll find the debug logs in the attached file...**

**Reply by pfolcato** [reply]
there is a key error in the inverter values, in the specific case it seems that the "p1_current" value does not exist....from what I understand...

**Comment by pfolcato** [comment]
try this script from shell: "python3 example.py 192.168.0.209 11502" and show me the result, this script print all the values of inverter, meter and battery, so I can check if all keys are presents
[example.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/10313486/example.zip)

**Comment by vinchz31** [comment]
Thanks to your script I was able to get most of the metrics to work by replacing the P's with L's (e.g. P1 => L1).
However, I can't find a key that matches ENERGY_CONSUMO_CASA, ACQUISTATA, etc.
I think that is where the problem lies.
However, I can get the total energy produced (via the Total Energy key) and what comes from or is sent to the power grid (via the Import tot (-) key). I have adapted the script to have these two values only, I still have to create a counter to add the two and that will give me the energy consumed by the house, unless you find in the keys a better way.

Thank you!

**Reply by pfolcato** [reply]
Energy_Home_Consumption (energy_consumo_casa),purchased (acquistata), Export (esportata) are keys that I calculated and create inside the script, after that I append the new key and the value in the  valuebattery dictionary...after that i update the Valuemeter dictionary and after that I update Invert_values... so you don't find in the inverter, but you need to modify Valuebattery.update (acquistata) ecc... to Valuemeter.update (acquistata) and so on.... and change the languages for better understanding :)

**Comment by vinchz31** [comment]
Thanks for your feedback.
I had a little trouble understanding but thanks to your help I understood how the script worked and the problems with the keys. I was able to retrieve the data I wanted by simplifying the script for my use and to create a meter allowing me to see the consumption of the house (in addition to the energy produced and that imported or redistributed on the network).
I think I'll adapt the script a bit more and post it here if it can help other people (even if I'm not sure my code is of high quality ;-) )
Many thanks for your help and the time you spent!
